### ISSUE TITLE
Better way than array_merge for Kohana_Database_Insert values()?

### ISSUE THREAD

**Issue by JackEllis** [kohana/core#600]
Hope everyone is doing well.

I'm wondering if we can come up with a better way of combining values than to use array_merge(). Obviously one might argue that the query builder should not be used when inserting 30,000+ rows at once, but it definitely could be if we sorted the values() method.

Any thoughts? Or is the general consensus "Don't use query builder for inserting large data sets"?

I ended up extracting the code from values() and keeping the query in a string, meaning I could avoid array_merge().

FYI - I can't find a profiling site that allows me to use such a large data set to prove my case, but here is some info below

9,999 BASIC arrays being added to ->values():

With array_merge / func_get_args() = 5.6 Seconds
Using foreach and $this->_values[] = $values; = 0.14 Seconds

Now obviously you can imagine how that scales up when you're inserting user profile rows which might have a lot of data.

Current

``` php
class Database_Query_Builder {
    private $_values = array();

    public function values(array $values)
    {
        if ( ! is_array($this->_values))
        {
            throw new Kohana_Exception('INSERT INTO ... SELECT statements cannot be combined with INSERT INTO ... VALUES');
        }

        // Get all of the passed values
        $values = func_get_args();

        $this->_values = array_merge($this->_values, $values);

        return $this;
    }   
}
```

Proposed

``` php
<?php
class Database_Query_Builder {
    private $_values = array();

    public function values(array $values)
    {
        if ( ! is_array($this->_values))
        {
            throw new Kohana_Exception('INSERT INTO ... SELECT statements cannot be combined with INSERT INTO ... VALUES');
        }

        // Get all of the passed values
        $values = func_get_args();

        foreach ($values as $entry)
        {
            $this->_values[] = $entry;   
        }

        return $this;
    }   
}
```

**Comment by JackEllis** [comment]
cc @enov

**Comment by acoulton** [comment]
I'm quite surprised by this, can you show us a sample of the calling code? I don't really understand why we're doing func_get_args either - it's a long time since I've used the database query builder so can't remember the usage.

**Comment by enov** [comment]
Hey @JackEllis! Yeah, a little search and looks like `array_merge` is [slow](http://www.bitbybit.dk/carsten/blog/?p=203), [affecting other frameworks](https://github.com/zendframework/zf2/issues/5716), and this is [solved in PHP 5.5.6](http://php.net/ChangeLog-5.php#5.5.6).

Need to have a deeper look if a `foreach` will be a good replacement of `array_merge` here for all cases.

**Comment by enov** [comment]
Ooops @acoulton did not notice you replied already.

**Comment by acoulton** [comment]
@enov no worries, thanks for the links - I never knew that.

**Comment by JackEllis** [comment]
@acoulton func_get_args() is there to support:

```
$insert->values($array1, $array2, $array3)
```

rather than forcing the user to do

```
$insert->values($array1);
$insert->values($array2);
$insert->values($array3);
```

@enov I had no idea that it has been fixed in 5.5.6! That's great news. I guess it depends on the minimum requirement for Kohana.

Here are some _rough_ tests I did on my localhost with 9,999 sets of small data.

**PHP 5.3.27**
With array_merge: 29.04 seconds
With foreach: 0.08 seconds

**PHP 5.4.19**
With array_merge: 26.23 seconds
With foreach: 0.07 seconds

**PHP 5.5.3**
With array_merge: 27.50 seconds
With foreach: 0.07 seconds

So basically, even with 5.5.6's fix, it's still going to make more sense to use foreach. Inserting 9,999 sets of data might be a rare use case, but the performance 'hit' on the smaller data sets (< 100 at a time) will be insignificant. The only case where using foreach isn't better would be 5.5.6+
